File: fabric-audit-agent/core/anomaly.js
function stats(ys) {
  const n = ys.length;
  const mean = ys.reduce((a, b) => a + b, 0) / n;
  const variance = ys.reduce((a, b) => a + (b - mean) ** 2, 0) / n;
  return { mean, stddev: Math.sqrt(variance) };
}

/**
 * Flag metrics that deviate > z standard deviations from their historical baseline. Pure.
 * @param {object} facts  current facts (reads facts.capacity.peakCuPct)
 * @param {Array<{metrics?:{peakCuPct?:number}}>} history  chronological run records
 * @param {{ z?:number, minPoints?:number }} [opts]
 * @returns {Array<{metric:string, resource:string, current:number, mean:number, stddev:number, sigma:number, direction:string, message:string}>}
 */
export function detectAnomalies(facts = {}, history = [], { z = 2, minPoints = 4 } = {}) {
  const anomalies = [];
  const series = history.map(h => h?.metrics?.peakCuPct).filter(v => typeof v === 'number');
  const current = facts?.capacity?.peakCuPct;

  if (series.length >= minPoints && typeof current === 'number') {
    const { mean, stddev } = stats(series);
    if (stddev > 0 && Math.abs(current - mean) > z * stddev) {
      const sigma = Math.round(((current - mean) / stddev) * 10) / 10;
      anomalies.push({
        metric: 'peakCuPct',
        resource: `capacity ${facts.capacity?.capacityId ?? ''}`.trim(),
        current,
        mean: Math.round(mean * 10) / 10,
        stddev: Math.round(stddev * 10) / 10,
        sigma,
        direction: current > mean ? 'above' : 'below',
        message: `Peak CU ${current}% is anomalous vs baseline (mean ${Math.round(mean)}%, ${Math.abs(sigma)}σ ${current > mean ? 'above' : 'below'}).`,
      });
    }
  }
  return anomalies;
}

File: fabric-audit-agent-py/fabric_audit_agent/anomaly.py
"""Statistical anomaly detection on the peak-CU series. Port of ``core/anomaly.js``. Pure."""
import math


def _is_num(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _r1(x):
    return math.floor(x * 10 + 0.5) / 10   # JS Math.round(x*10)/10 (half-up, works for negatives)


def _fmt(x):
    return str(int(x)) if x == int(x) else str(x)


def _stats(ys):
    n = len(ys)
    mean = sum(ys) / n
    variance = sum((b - mean) ** 2 for b in ys) / n
    return mean, math.sqrt(variance)


def detect_anomalies(facts=None, history=None, z=2, min_points=4):
    facts = facts or {}
    history = history or []
    anomalies = []
    series = [v for v in ((h.get("metrics") or {}).get("peakCuPct") for h in history if isinstance(h, dict)) if _is_num(v)]
    current = (facts.get("capacity") or {}).get("peakCuPct")

    if len(series) >= min_points and _is_num(current):
        mean, stddev = _stats(series)
        if stddev > 0 and abs(current - mean) > z * stddev:
            sigma = _r1((current - mean) / stddev)
            direction = "above" if current > mean else "below"
            anomalies.append({
                "metric": "peakCuPct",
                "resource": f"capacity {(facts.get('capacity') or {}).get('capacityId') or ''}".strip(),
                "current": current,
                "mean": _r1(mean),
                "stddev": _r1(stddev),
                "sigma": sigma,
                "direction": direction,
                "message": f"Peak CU {_fmt(current)}% is anomalous vs baseline (mean {int(math.floor(mean + 0.5))}%, {_fmt(abs(sigma))}σ {direction}).",
            })
    return anomalies
